File: reports/nato-members.py
"""
Demo script using (Py-) MuPDF "Story" feature.

"""
import zipfile
from pprint import pprint
from Reports import LongTable, FrontMatterReport, SimpleBlock
import fitz

HEADER = """<h2 style="text-align:center;font-family: sans-serif;">Members of the NATO</h2>"""
HTML = """
<style>
body {font-family: sans-serif;font-size: 14px;}
td, th {
    padding-left: 10px;
    padding-right: 10px;
}
.even {background-color: #ccc;} /* use for alternating background */
.odd {background-color: #fff;} /* use for alternating background */

</style>

<body>
<table>
<tr id="header">
    <th>Country</th>
    <th>Type</th>
    <th>Flag</th>
    <th>Since</th>
</tr>

<tr id="template">
    <td id="country"></td>
    <td id="member"></td>
    <td id="flag"></td>
    <td id="since"></td>
</tr>
</table>
</body>
"""

national_flags = zipfile.ZipFile("national-flags.zip")


def fetch_rows():
    table_data = """country;member;flag;since
    Belgium;Founder;|img|Belgium.jpg;1949
    Denmark;Founder;|img|Denmark.jpg;1949
    France;Founder;|img|France.jpg;1949
    Iceland;Founder;|img|Iceland.jpg;1949
Italy;Founder;|img|Italy.jpg;1949
Canada;Founder;|img|Canada.jpg;1949
Luxembourg;Founder;|img|Luxembourg.jpg;1949
Netherlands;Founder;|img|Netherlands.jpg;1949
Norway;Founder;|img|Norway.jpg;1949
Portugal;Founder;|img|Portugal.jpg;1949
United Kingdom;Founder;|img|United_Kingdom.jpg;1949
United States;Founder;|img|United_States.jpg;1949
Greece;Joiner;|img|Greece.jpg;1952
Turkey;Joiner;|img|Turkey.jpg;1952
Germany;Joiner;|img|Germany.jpg;1955
Spain;Joiner;|img|Spain.jpg;1982
Poland;Joiner;|img|Poland.jpg;1999
Czech Republic;Joiner;|img|Czech_Republic.jpg;1999
Hungary;Joiner;|img|Hungary.jpg;1999
Bulgaria;Joiner;|img|Bulgaria.jpg;2004
Estonia;Joiner;|img|Estonia.jpg;2004
Latvia;Joiner;|img|Latvia.jpg;2004
Lithuania;Joiner;|img|Lithuania.jpg;2004
Romania;Joiner;|img|Romania.jpg;2004
Slovakia;Joiner;|img|Slovakia.jpg;2004
Slovenia;Joiner;|img|Slovenia.jpg;2004
Albania;Joiner;|img|Albania.jpg;2009
Croatia;Joiner;|img|Croatia.jpg;2009
Montenegro;Joiner;|img|Montenegro.jpg;2017
North Macedonia;Joiner;|img|North_Macedonia.jpg;2020
Finland;Joiner;|img|Finland.jpg;2023
Sweden;Joiner;|img|Sweden.jpg;2023
"""
    data = [l.split(";") for l in table_data.splitlines()]
    return data


report = FrontMatterReport(fitz.paper_rect("a5"))
header = SimpleBlock(html=HEADER, report=report)
items = LongTable(
    report=report,
    html=HTML,
    top_row="header",
    top_row_bg="#ffff00",
    fetch_rows=fetch_rows,
    archive=national_flags,
    alternating_bg=("#c0c0c0", "#ffffff"),
)

report.pages = report.page0 = [header, items]
report.run("NATO.pdf")

File: itext/nato_members/src/main/java/org/example/NatoMembers.java
package org.example;

import com.itextpdf.io.image.ImageData;
import com.itextpdf.io.image.ImageDataFactory;
import com.itextpdf.kernel.colors.Color;
import com.itextpdf.kernel.colors.ColorConstants;
import com.itextpdf.kernel.geom.PageSize;
import com.itextpdf.kernel.pdf.PdfDocument;
import com.itextpdf.kernel.pdf.PdfWriter;
import com.itextpdf.layout.Document;
import com.itextpdf.layout.properties.TextAlignment;
import com.itextpdf.layout.element.Cell;
import com.itextpdf.layout.element.Image;
import com.itextpdf.layout.element.Paragraph;
import com.itextpdf.layout.element.Table;
import com.itextpdf.layout.properties.HorizontalAlignment;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

public class NatoMembers {
    public static String data = "Belgium;Founder;Belgium.jpg;1949\n" +
            "Denmark;Founder;Denmark.jpg;1949\n" +
            "France;Founder;France.jpg;1949\n" +
            "Iceland;Founder;Iceland.jpg;1949\n" +
            "Italy;Founder;Italy.jpg;1949\n" +
            "Canada;Founder;Canada.jpg;1949\n" +
            "Luxembourg;Founder;Luxembourg.jpg;1949\n" +
            "Netherlands;Founder;Netherlands.jpg;1949\n" +
            "Norway;Founder;Norway.jpg;1949\n" +
            "Portugal;Founder;Portugal.jpg;1949\n" +
            "United Kingdom;Founder;United_Kingdom.jpg;1949\n" +
            "United States;Founder;United_States.jpg;1949\n" +
            "Greece;Joiner;Greece.jpg;1952\n" +
            "Turkey;Joiner;Turkey.jpg;1952\n" +
            "Germany;Joiner;Germany.jpg;1955\n" +
            "Spain;Joiner;Spain.jpg;1982\n" +
            "Poland;Joiner;Poland.jpg;1999\n" +
            "Czech Republic;Joiner;Czech_Republic.jpg;1999\n" +
            "Hungary;Joiner;Hungary.jpg;1999\n" +
            "Bulgaria;Joiner;Bulgaria.jpg;2004\n" +
            "Estonia;Joiner;Estonia.jpg;2004\n" +
            "Latvia;Joiner;Latvia.jpg;2004\n" +
            "Lithuania;Joiner;Lithuania.jpg;2004\n" +
            "Romania;Joiner;Romania.jpg;2004\n" +
            "Slovakia;Joiner;Slovakia.jpg;2004\n" +
            "Slovenia;Joiner;Slovenia.jpg;2004\n" +
            "Albania;Joiner;Albania.jpg;2009\n" +
            "Croatia;Joiner;Croatia.jpg;2009\n" +
            "Montenegro;Joiner;Montenegro.jpg;2017\n" +
            "North Macedonia;Joiner;North_Macedonia.jpg;2020\n" +
            "Finland;Joiner;Finland.jpg;2023\n" +
            "Sweden;Observer;Sweden.jpg;N/A";

    public static void main(String[] args) throws IOException {
        Color backColor;
        float tableWidth = 320f;
        float[] pointColumnWidths = {200f, 150f, 50f, 50f};
        String[] columns = {"Country", "Type", "Flag", "Since"};
        int numColumn = 4;

        ZipFile zipFile = new ZipFile("src/main/resources/national-flags.zip");

        PdfWriter writer = new PdfWriter("nato_members.pdf");

        // Creating a PdfDocument
        PdfDocument pdfDoc = new PdfDocument(writer);

        // Creating a Document
        Document document = new Document(pdfDoc, PageSize.LETTER, false);

        // Adding header
        Paragraph header = new Paragraph("Members of the NATO");
        header.setTextAlignment(TextAlignment.CENTER);
        header.setFontSize(20);
        header.setBold();

        document.add(header);

        // Initialize table
        Table table = new Table(pointColumnWidths);
        table.setWidth(tableWidth);
        table.setHorizontalAlignment(HorizontalAlignment.CENTER);

        // Adding header cells to the table
        for (int i =0; i <numColumn ; i ++) {
            table.addHeaderCell(new Cell().add(new Paragraph(columns[i])).setBold().setBorder(null));
        }

        // Adding cells to the table
        String[] rows = data.split("\n");

        for (int i =0; i < rows.length; i++) {
            if (i % 2 == 0)
                backColor = ColorConstants.LIGHT_GRAY;
            else
                backColor = ColorConstants.WHITE;

            String[] values = rows[i].split(";");

            ZipEntry entry = zipFile.getEntry(values[2]);

            if(entry != null) {
                InputStream stream = zipFile.getInputStream(entry);

                // Read the stream and then close it
                byte[] bytes = readStream(stream);
                stream.close();

                // Convert byte array to ImageData
                ImageData imageData = ImageDataFactory.create(bytes);

                // Now, use imageData object wherever you need
                for (int j = 0; j < values.length; j++) {
                    if (j == 2)
                        table.addCell(new Cell().add(new Image(imageData)).setBackgroundColor(backColor).setBorder(null));
                    else
                        table.addCell(new Cell().add(new Paragraph(values[j])).setBackgroundColor(backColor).setBorder(null));
                }
            }
        }

        document.add(table);

        document.close();
    }

    public static byte[] readStream(InputStream input) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        byte[] buffer = new byte[4096];
        int n;
        while ((n = input.read(buffer)) != -1) {
            output.write(buffer, 0, n);
        }
        return output.toByteArray();
    }
}
